python-oauth-web-app/app.py

```diff
@@ -0,0 +1,336 @@
+        authsrv_url, authsrv_meta, did, app_url, secret_jwk
+    )
+    resp.raise_for_status()
+    # This field is confusingly named: it is basically a token to refering back to the successful PAR request.
+    par_request_uri = resp.json()["request_uri"]
+
+    print(f"saving oauth_auth_request to DB: {state}")
+    query_db(
+        "INSERT INTO oauth_auth_request (state, iss, nonce, pkce_verifier, dpop_private_jwk, did, username, pds_url) VALUES(?, ?, ?, ?, ?, ?, ?, ?);",
+        [
+            state,
+            authsrv_meta["issuer"],
+            nonce,
+            pkce_verifier,
+            dpop_private_jwk,
+            did,
+            username,
+            pds_url,
+        ],
+    )
+
+    # Forward the user to the Authorization Server to complete the browser auth flow.
+    # IMPORTANT: Authorization endpoint URL is untrusted input, security mitigations are needed before redirecting user
+    auth_url = authsrv_meta["authorization_endpoint"]
+    assert is_safe_url(auth_url)
+    qparam = urlencode({"client_id": client_id, "request_uri": par_request_uri})
+    return redirect(f"{auth_url}?{qparam}")
+
+
+# Endpoint for receiving "callback" responses from the Authorization Server, to complete the auth flow.
+@app.route("/oauth/callback")
+def oauth_callback():
+    state = request.args["state"]
+    iss = request.args["iss"]
+    code = request.args["code"]
+
+    # Lookup auth request by the "state" token (which we randomly generated earlier)
+    row = query_db(
+        "SELECT * FROM oauth_auth_request WHERE state = ?;",
+        [state],
+        one=True,
+    )
+    if row is None:
+        abort(400, "OAuth request not found")
+
+    # Delete row to prevent response replay
+    query_db("DELETE FROM oauth_auth_request WHERE state = ?;", [state])
+
+    # Verify query param "iss" against earlier oauth request "iss"
+    assert row["iss"] == iss
+    # This is redundant with the above SQL query, but also double-checking that the "state" param matches the original request
+    assert row["state"] == state
+
+    # Complete the auth flow by requesting auth tokens from the authorization server.
+    app_url = request.url_root.replace("http://", "https://")
+    tokens, dpop_nonce = complete_auth_request(row, code, app_url, secret_jwk)
+
+    # Save session (including auth tokens) in database
+    query_db(
+        "INSERT OR REPLACE INTO oauth_session (did, username, pds_url, iss, access_token, refresh_token, dpop_nonce, dpop_private_jwk) VALUES(?, ?, ?, ?, ?, ?, ?, ?);",
+        [
+            row["did"],
+            row["username"],
+            row["pds_url"],
+            row["iss"],
+            tokens["access_token"],
+            tokens["refresh_token"],
+            dpop_nonce,
+            row["dpop_private_jwk"],
+        ],
+    )
+
+    # Set a (secure) session cookie in the user's browser, for authentication between the browser and this app
+    session["user_did"] = row["did"]
```

Comment:
If all that's stored in cookies is a did, can I start using someone's session simply by knowing their did?

Reply:
The `FLASK_SECRET_KEY` environment variable is used to encrypt session content.

See https://flask.palletsprojects.com/en/2.3.x/config/#SECRET_KEY